python/paddle/ir/backward.py

```diff
@@ -0,0 +1,640 @@
+            )
+
+
+def update_no_grad_set_stopgradient(block, no_grad_set):
+    for op in block.ops:
+        for opresult_idx in range(op.num_results()):
+            value = op.result(opresult_idx)
+            if value.stop_gradient and value not in no_grad_set:
+                no_grad_set.add(value)
+
+
+def update_all_structure(backward_ops, op_to_opgrad_list, gradop):
+    backward_ops.append(gradop)
+    op_to_opgrad_list.append(gradop)
+
+
+def prepare_grad_outputs(
+    block, grad_outputs, outputs, value_to_valuegrad, op_to_opgrad
+):
+    """
+    if grad_outputs is none, add fill_1 op to create grad_outputs
+    else check if outputs shape and dtype is same to grad_outputs, else raise error
+
+    if only part of op's outputs in outputs, add fill_0 op to create other grad_outputs
+    eg: split
+
+    update value_to_valuegrad and op_to_opgrad
+    return complete_outputs and complete_gradoutputs, backward_ops
+
+    """
+    if not grad_outputs:
+        grad_outputs = [None] * len(outputs)
+
+    if len(grad_outputs) != len(outputs):
+        raise ValueError(
+            "Should have the same number of grad_outputs as outputs"
+        )
+    backward_ops = []
+    for i, grad in enumerate(grad_outputs):
+        output = outputs[i]
+        # fwd : op1 -> op2 -> op3 -> output
+        # bwd : op1G <- op2G <- op3G <- outputG <- fillop/feedop
+        if grad is None:
+            paddle.full(
+                output.shape,
+                1.0,
+                dtype=output.dtype,
+            )
+            op_list = block.ops
+            fillop = op_list[len(op_list) - 1]
+
+            update_all_structure(
+                backward_ops,
+                op_to_opgrad[output.get_defining_op()],
+                fillop,
+            )
+            value_to_valuegrad[output] = [[fillop.result(0)]]
+        else:
+            if output.shape != grad.shape:
+                raise ValueError(
+                    "The shape of grad_output[%d] should be the same as the shape of output[%d]"
+                    % (i, i)
+                )
+            if output.dtype != grad.dtype:
+                raise ValueError(
+                    "The dtype of grad_output[%d] should be the same as the dtype of output[%d]"
+                    % (i, i)
+                )
+            feedop = grad.get_defining_op()
+            update_all_structure(
+                backward_ops, op_to_opgrad[output.get_defining_op()], feedop
+            )
+            value_to_valuegrad[output] = [[grad]]
+
+    # add input for bwd first op
+    complete_outputs = outputs
+    complete_gradoutputs = grad_outputs
+
+    visited_output = set()
+    for output in outputs:
+        if output in visited_output:
+            continue
+        for i in range(output.get_defining_op().num_results()):
+            opresult = output.get_defining_op().result(i)
+            if opresult in value_to_valuegrad.keys():
```

Comment:
```suggestion
            if opresult in value_to_valuegrad:
```
value_to_valuegrad 是个map吧，直接查找应该是hash形式，O(1)?，keys() 返回的是迭代器，再查找的话就是顺序遍历，O(N) ？